crates/bevy_tasks/src/task_pool.rs

```diff
@@ -267,41 +277,28 @@ impl TaskPool {
         if spawned.is_empty() {
             Vec::new()
         } else {
-            let get_results = async move {
-                let mut results = Vec::with_capacity(spawned.len());
-                while let Ok(task) = spawned.pop() {
-                    results.push(task.await);
-                }
+            future::block_on(async move {
+                let get_results = async move {
+                    let mut results = Vec::with_capacity(scope.spawned.len());
+                    while let Ok(task) = spawned.pop() {
+                        results.push(task.await);
+                    }
 
-                results
-            };
-
-            // Pin the futures on the stack.
-            pin!(get_results);
-
-            // SAFETY: This function blocks until all futures complete, so we do not read/write
-            // the data from futures outside of the 'scope lifetime. However,
-            // rust has no way of knowing this so we must convert to 'static
-            // here to appease the compiler as it is unable to validate safety.
-            let get_results: Pin<&mut (dyn Future<Output = Vec<T>> + 'static + Send)> = get_results;
-            let get_results: Pin<&'static mut (dyn Future<Output = Vec<T>> + 'static + Send)> =
-                unsafe { mem::transmute(get_results) };
-
-            // The thread that calls scope() will participate in driving tasks in the pool
-            // forward until the tasks that are spawned by this scope() call
-            // complete. (If the caller of scope() happens to be a thread in
-            // this thread pool, and we only have one thread in the pool, then
-            // simply calling future::block_on(spawned) would deadlock.)
-            let mut spawned = task_scope_executor.spawn(get_results);
-
-            loop {
-                if let Some(result) = future::block_on(future::poll_once(&mut spawned)) {
-                    break result;
+                    results
                 };
 
-                self.executor.try_tick();
```

Comment:
This will deadlock `scope` if all threads are running one simultaneously. Perhaps a dedicated render thread instead?

Reply:
agree this is a problem. Separate render thread doesn't help anything as the render stages still run a scope and will still run systems from the main stages. Ideally scope would only run tasks that are spawned on it, but probably not possible with async_executor. The only other viable solution I can think of is to separate out the tasks pools between the main thread and the render thread, which will cause thread contention between the pools, but maybe that is better than this solution.

Reply:
I think the *ideal* solution would be removing `!Send` resources.

If we never had to move worlds back onto the main (`winit`) thread, we could just have three dedicated threads⁠—`winit`, app, and render⁠—and never deal with this again. 

The solution I envision involves a non-`async` version of `ThreadExecutor` and some `T` that stores all the `!Send` resources. `T` is stored on `App` until the app starts, then it's left on the main thread. The main thread would own a `ThreadExecutor` and process it during `MainEventsCleared` (or whatever). The other threads would send functions/closures that access `T` to the executor through a channel and block until they receive a result.

With that, we could just wrap the parts of `create_windows` and `prepare_windows` that access `!Send` data. 

The executor wouldn't even need `async`.
With dedicated threads, nothing blocks the `winit` event loop. It could run at full speed.

Reply:
would another priority in https://github.com/bevyengine/bevy/pull/4740 help here? Can we only run tasks of a certain priority in the scope with the new executor there?

Reply:
> I think the ideal solution would be removing !Send resources.

This issue has nothing to do with !Send resources. It has to do with systems from the main app running on the render thread per this comment https://github.com/bevyengine/bevy/pull/6503#issuecomment-1309791442

Reply:
I guess it's a general comment, but I commented on this thread because the first comment suggested a dedicated render thread, which imo is where we'll ultimately end up.

By "this issue" I'm assuming you mean this particular thread, since `!Send` resources are relevant to this PR. If there were no `!Send` resources, this implementation would be a lot simpler.

Reply:
going to add another task pool to fix this. not the ideal fix, but only reasonable one without major changes to how task pools work